### PR TITLE
V1.3: Find enhancements

### PR DESCRIPTION
**Pull request by filbertphang**
- Implement find-by-tag
- Implement find-by-ingredient

### CODE REVIEW

File: src/main/java/seedu/recipe/commons/core/Messages.java [hunk 1]
```diff
@@ -7,6 +7,7 @@ public class Messages {
 
     public static final String MESSAGE_UNKNOWN_COMMAND = "Unknown command";
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
+    public static final String MESSAGE_EMPTY_KEYWORDS_FIND = "Keywords for find cannot be empty!";
```
Comment (SeeuSim) [inline src/main/java/seedu/recipe/commons/core/Messages.java:10]: minor nitpick but `"Keywords for the 'find' command ..."` might be better

File: src/main/java/seedu/recipe/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -23,12 +31,49 @@ public FindCommand parse(String args) throws ParseException {
         String trimmedArgs = args.trim();
         if (trimmedArgs.isEmpty()) {
             throw new ParseException(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+                String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
 
-        String[] nameKeywords = trimmedArgs.split("\\s+");
+        String[] allKeywords = trimmedArgs.split("\\s+");
```
Comment (SeeuSim) [inline src/main/java/seedu/recipe/logic/parser/FindCommandParser.java:37]: This should work for space separated keywords too, smart

File: src/main/java/seedu/recipe/model/recipe/PropertyNameContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * Constructs an instance of PropertyNameContainsKeywordsPredicate.
+     *
+     * @param keywords       A list of keywords to test against the property name
+     * @param propertyGetter A method that returns the property from a Recipe
+     * @param nameGetter     A method that gets the name of a property
+     */
+    public PropertyNameContainsKeywordsPredicate(List<String> keywords, Function<Recipe, T> propertyGetter,
+                                                 Function<T, String> nameGetter) {
```
Comment (SeeuSim) [inline src/main/java/seedu/recipe/model/recipe/PropertyNameContainsKeywordsPredicate.java:33]: Personal preference but line continuation might be better if the second line is at indentation level +4 wrt function body

```java
    methodName(...
            paramX, ... ) {
        //Statement 1 of function
```

**Review by SeeuSim — APPROVED** [review]
Code quality and functionality looks good, and great to see that extensive unit tests were added

File: src/main/java/seedu/recipe/model/recipe/PropertyNameContainsKeywordsPredicate.java (hunk 3 above)
Comment (filbertphang) [reply src/main/java/seedu/recipe/model/recipe/PropertyNameContainsKeywordsPredicate.java:33]: there doesn't seem to be an option for this under intelliJ code style, so i'm leaving it as-is

**Pull request merged**